<launch>

  <arg name="rviz" default="true"/>
  <arg name="launch_manager" default="true"/>
  <arg name="manager" default="mirror_recognition_manager"/>

  <group if="$(arg launch_manager)">
    <node name="$(arg manager)"
          pkg="nodelet" type="nodelet"
          args="manager"
          output="screen" respawn="true"/>
  </group>

  <group ns="kinect_head_remote">
    <!-- Relay -->
    <node name="relay_camera_info"
          pkg="topic_tools" type="relay"
          args="/kinect_head/rgb/camera_info /kinect_head_remote/rgb/camera_info"
          output="screen" respawn="true"/>
    <!-- Republish images -->
    <node name="republish_rgb"
          pkg="image_transport" type="republish"
          args="compressed raw"
          output="screen" respawn="true">
      <remap from="in" to="/kinect_head/rgb/image_rect_color"/>
      <remap from="out" to="rgb/image_rect_color"/>
    </node>
    <node name="republish_depth"
          pkg="image_transport" type="republish"
          args="compressedDepth raw"
          output="screen" respawn="true">
      <remap from="in" to="/kinect_head/depth_registered/hw_registered/image_rect"/>
      <remap from="out" to="depth_registered/hw_registered/image_rect"/>
    </node>
    <!-- 3D reconstruction -->
    <node name="point_cloud_xyzrgb_remote"
          pkg="nodelet" type="nodelet"
          args="load depth_image_proc/point_cloud_xyzrgb /$(arg manager)"
          output="screen" respawn="true">
      <remap from="rgb/camera_info" to="rgb/camera_info"/>
      <remap from="rgb/image_rect_color" to="rgb/image_rect_color"/>
      <remap from="depth_registered/image_rect" to="depth_registered/hw_registered/image_rect"/>
    </node>
  </group>

  <!-- Assemble /tilt_scan. -->
  <!-- The point cloud is published only when service is called. -->
  <node name="tilt_scan_assembler"
        pkg="laser_assembler" type="laser_scan_assembler"
        output="screen" respawn="true">
    <remap from="scan" to="/tilt_scan"/>
    <remap from="assemble_scans2" to="assemble_tilt_scan"/>
    <rosparam>
      fixed_frame: laser_tilt_link
      max_scans: 400
    </rosparam>
  </node>

  <!-- Always publish assembled scans as timer -->
  <node name="assembled_scans_timer_publisher"
        pkg="mirror_recognition" type="assembled_scans_timer_publisher.py"
        output="screen"
        respawn="true">
    <remap from="~assemble_scans2" to="assemble_tilt_scan"/>
    <rosparam>
      rate: 40
      duration: 2.0
    </rosparam>
  </node>

  <!-- Main Process -->
  <include file="$(find mirror_recognition)/launch/include/point_cloud_mirror_flipper.launch">
    <arg name="INPUT_IMAGE" value="/kinect_head_remote/rgb/image_rect_color"/>
    <arg name="INPUT_INFO" value="/kinect_head_remote/rgb/camera_info"/>
    <arg name="INPUT_DEPTH" value="/kinect_head_remote/depth_registered/hw_registered/image_rect"/>
    <!-- <arg name="INPUT_CLOUD" value="/kinect_head_remote/depth_registered/points"/> -->
    <arg name="INPUT_CLOUD" value="assembled_scans_timer_publisher/output"/>
    <arg name="LAUNCH_MANAGER" value="false"/>
    <arg name="manager" value="$(arg manager)"/>
    <arg name="fcn_at_once_concat_at_once" value="true"/>
    <arg name="model_dir" value="$(find mirror_recognition)/logs/20190127-094644"/>
    <arg name="gpu" value="0"/>
  </include>

  <!-- Visualization -->
  <!-- Visualize input depth image -->
  <group ns="kinect_head_remote/depth_registered/hw_registered">
    <node name="image_rect_viz"
          pkg="image_view" type="image_view"
          output="screen"
          respawn="true">
      <remap from="image" to="image_rect"/>
      <rosparam>
        gui: false
        colormap: 2
      </rosparam>
    </node>
  </group>

  <!-- Visualize output depth image -->
  <group ns="fcn_mirror_segmentation_depth_estimation/output">
    <node name="depth_inpainted_viz"
          pkg="image_view" type="image_view"
          output="screen"
          respawn="true">
      <remap from="image" to="depth_inpainted"/>
      <rosparam>
        gui: false
        colormap: 2
      </rosparam>
    </node>
    <node name="depth_pred_raw_viz"
          pkg="image_view" type="image_view"
          output="screen"
          respawn="true">
      <remap from="image" to="depth_pred_raw"/>
      <rosparam>
        gui: false
        colormap: 2
      </rosparam>
    </node>
    <node name="depth_pred_labeled_viz"
          pkg="image_view" type="image_view"
          output="screen"
          respawn="true">
      <remap from="image" to="depth_pred_labeled"/>
      <rosparam>
        gui: false
        colormap: 2
      </rosparam>
    </node>
  </group>

  <node name="rosconsole_overlay_text"
        pkg="jsk_rviz_plugins" type="rosconsole_overlay_text.py"
        output="screen"
        respawn="true">
    <rosparam>
      reverse_lines: false
    </rosparam>
  </node>

  <!-- RViz -->
  <group if="$(arg rviz)">
    <node name="rviz"
          pkg="rviz" type="rviz"
          args="-d $(find mirror_recognition)/rvizconfig/point_cloud_flipper_by_pr2_kinect.rviz"/>
  </group>

</launch>
